src/spartan/snark.rs

```diff
@@ -101,6 +101,26 @@ impl<G: Group, EE: EvaluationEngineTrait<G>> RelaxedR1CSSNARKTrait<G> for Relaxe
   ) -> Result<(Self::ProverKey, Self::VerifierKey), SpartanError> {
     let mut cs: ShapeCS<G> = ShapeCS::new();
     let _ = circuit.synthesize(&mut cs);
+
+    // Padding the ShapeCS: constraints (rows) and variables (columns)
+    let num_constraints = cs.num_constraints();
+
+    (num_constraints..num_constraints.next_power_of_two()).for_each(|i| {
+      cs.enforce(
+        || format!("padding_constraint_{i}"),
+        |lc| lc,
+        |lc| lc,
+        |lc| lc,
+      )
+    });
```

Comment:
I don't understand what this code does. Is this a dummy constraint?

Reply:
Precisely. Recall that in an R1CS the number of constraints corresponds to the number of rows. The issue with the previous implementation was that if that number was not a power of 2, the compiler would break (funnily enough, the only example they had, the `CubicCircuit`, had exactly 4 constraints by nature - although I think I have a way to do it with 3). So we add "empty constraints" where all coefficients are zero - and yes, that's what that slightly odd syntax is doing. We also needed to add zero-valued variables to pad the number of columns to a power of 2.

Reply:
Alright, makes sense on a high level. I'm not as familiar with the internals as you at this point, let's go upstream with this and see.